File: backend/src/ytforge/interfaces/api/schemas/trends.py
from __future__ import annotations

import uuid
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from ytforge.domain.enums import TrendSource


class TrendCreateRequest(BaseModel):
    source: TrendSource
    topic: str = Field(min_length=1, max_length=500)
    url: str | None = None
    score: float = Field(default=0.0, ge=0.0, le=100.0)
    raw_payload: dict[str, Any] = Field(default_factory=dict)


class TrendRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    channel_id: uuid.UUID | None
    source: TrendSource
    topic: str
    url: str | None
    score: float
    raw_payload: dict[str, Any]

File: frontend/src/lib/api/schemas/trends.ts
import { z } from "zod";
import { TrendSource } from "@/types/enums";

const trendSourceSchema = z.enum([
  TrendSource.GOOGLE_TRENDS,
  TrendSource.YOUTUBE_TRENDING,
  TrendSource.REDDIT,
  TrendSource.HACKER_NEWS,
  TrendSource.X,
  TrendSource.RSS,
  TrendSource.NEWS_API,
]);

export const trendReadSchema = z.object({
  id: z.string().uuid(),
  channel_id: z.string().uuid().nullable(),
  source: trendSourceSchema,
  topic: z.string(),
  url: z.string().nullable(),
  score: z.number(),
  raw_payload: z.record(z.string(), z.unknown()),
});
export type TrendRead = z.infer<typeof trendReadSchema>;

export type TrendCreateRequest = {
  source: TrendSource;
  topic: string;
  url?: string | null;
  score?: number;
  raw_payload?: Record<string, unknown>;
};
